File: laba_3/task_3/task_3.cpp
#include <iostream>
#include <cmath>
#include <cstdint>

using namespace std;

uint64_t nod(uint64_t x, uint64_t y) { // НОД
    while (y != 0) {
        x %= y;
        swap(x, y);
    }
    return x;
}

int main() {
    uint64_t power;  // степень в числителе: n^power
    uint64_t base;   // основание в знаменателе: base^n
    cin >> power >> base;

    if (base <= 1) { // если ряд расходится
        cout << "infinity" << endl;
        return 0;
    }

    double sum = 0; // накопленная сумма ряда
    double term = 0; // текущее слагаемое

    for (uint64_t n = 1; n <= 20000; n++) { // вычисляем сумму ряда

        double numerator = 1;
        for (uint64_t i = 0; i < power; i++) {  // Вычисляем n^power
            numerator *= n;
        }

        double denominator = 1;
        for (uint64_t i = 0; i < n; i++) {  // Вычисляем base^n
            denominator *= base;
        }

        term = numerator / denominator;
        sum += term;

        // Если слагаемое стало очень маленьким, останавливаемся
        if (term < 1e-18) {
            break;
        }
    }

    uint64_t bestNum = 1;
    uint64_t bestDen = 1;
    double bestDiff = 1e20;

    // Перебираем знаменатели, чтобы найти лучшую дробь
    for (uint64_t den = 1; den <= 2000000; den++) {
        // Находим ближайший числитель
        uint64_t num = (uint64_t)(sum * den + 0.5);

        // Проверяем насколько близко
        double diff = fabs((double)num / den - sum);

        if (diff < bestDiff) {
            bestDiff = diff;
            bestNum = num;
            bestDen = den;
            // Если нашли очень точное совпадение - можно остановиться
            if (diff < 1e-15) {
                break;
            }
        }
    }

    // Сокращаем дробь
    uint64_t common = nod(bestNum, bestDen);
    cout << bestNum / common << "/" << bestDen / common << endl;

    return 0;
}

File: laba_3/task_3/task_3.rs
//use std::cmp::Ordering;

fn nod(mut x: u64, mut y: u64) -> u64 {
    // НОД (алгоритм Евклида)
    while y != 0 {
        x %= y;
        std::mem::swap(&mut x, &mut y);
    }
    x
}

fn main() {
    let mut input = String::new();
    
    // Читаем power и base
    std::io::stdin().read_line(&mut input).unwrap();
    let mut parts = input.trim().split_whitespace();
    let power: u64 = parts.next().unwrap().parse().unwrap();
    let base: u64 = parts.next().unwrap().parse().unwrap();

    if base <= 1 {
        // если ряд расходится
        println!("infinity");
        return;
    }

    let mut sum: f64 = 0.0; // накопленная сумма ряда
    
    for n in 1..=20000u64 {
        // вычисляем сумму ряда
        
        let mut numerator: f64 = 1.0;
        for _ in 0..power {
            // Вычисляем n^power
            numerator *= n as f64;
        }

        let mut denominator: f64 = 1.0;
        for _ in 0..n {
            // Вычисляем base^n
            denominator *= base as f64;
        }

        let term = numerator / denominator;
        sum += term;

        // Если слагаемое стало очень маленьким, останавливаемся
        if term < 1e-18 {
            break;
        }
    }

    let mut best_num: u64 = 1;
    let mut best_den: u64 = 1;
    let mut best_diff: f64 = 1e20;

    // Перебираем знаменатели, чтобы найти лучшую дробь
    for den in 1..=2000000u64 {
        // Находим ближайший числитель
        let num = (sum * den as f64 + 0.5) as u64;

        // Проверяем насколько близко
        let diff = ((num as f64 / den as f64) - sum).abs();

        if diff < best_diff {
            best_diff = diff;
            best_num = num;
            best_den = den;
            
            // Если нашли очень точное совпадение - можно остановиться
            if diff < 1e-15 {
                break;
            }
        }
    }

    // Сокращаем дробь
    let common = nod(best_num, best_den);
    println!("{}/{}", best_num / common, best_den / common);
}